packages/keystone/authStrategy/passwordAuth.js

```diff
@@ -0,0 +1,131 @@
+ *     secretField: 'password',
+ *     protectIdentities: true,
+ *
+ *     // custom props
+ *     itemTypeField: 'type',
+ *     itemTypeOptions: ['staff', 'service', 'resident'],
+ *     itemTypeDefault: 'staff',
+ *
+ *     verificationField: 'isEmailVerified',
+ *
+ *     rejectSoftDeletedItems: true,
+ * }
+ */
+class ExtendedPasswordAuthStrategy extends DefaultPasswordAuthStrategy {
+    constructor (...props) {
+        super(...props)
+
+        this.config = {
+            // NOTE: We do not set default values for new fields as the implementation of our applications may differ.
+            // We should configure each application yourself as we need.
+            //
+            // itemTypeField: 'type',
+            // itemTypeOptions: ['staff', 'service', 'resident'],
+            // itemTypeDefault: 'staff',
+            //
+            // verificationField: 'isEmailVerified',
+
+            rejectSoftDeletedItems: true,
+
+            ...this.config,
+        }
+    }
+
+    getInputFragment () {
+        const itemTypeField = this.config.itemTypeField
+            ? `${this.config.itemTypeField}: String`
+            : ''
+
+        return `
+          ${this.config.identityField}: String
+          ${this.config.secretField}: String
+          ${itemTypeField}
+        `
+    }
+
+    async validate (args) {
+        const { itemTypeField, itemTypeOptions, itemTypeDefault } = this.config
+        const itemType = args[itemTypeField] || itemTypeDefault || null
+
+        if (Boolean(itemTypeField) && Array.isArray(itemTypeOptions) && !itemTypeOptions.includes(itemType)) {
+            const message = `[passwordAuth:itemType:invalid] The ${itemTypeField} contains an invalid value. It must be one of the following: ${itemTypeOptions.join(', ')}`
+            return { success: false, message }
+        }
+
+        return await super.validate(args)
+    }
+
+    async _getItem (list, args, secretFieldInstance) {
+        // Match by identity
+        const {
+            identityField,
+            itemTypeField,
+            itemTypeOptions,
+            itemTypeDefault,
+            verificationField,
+            rejectSoftDeletedItems,
+        } = this.config
+
+        const identity = args[identityField]
+        const itemType = args[itemTypeField] || itemTypeDefault || null
+        const results = await list.adapter.find({
+            [identityField]: identity,
+            ...(itemTypeField && Array.isArray(itemTypeOptions) ? { [itemTypeField]: itemType } : null),
+            ...(rejectSoftDeletedItems ? { deletedAt: null } : null),
+            ...(verificationField ? { [verificationField]: true } : null),
+        })
+
+        // * * * Next comes the original logic * * *
+        //
+        // If we failed to match an identity and we're protecting existing identities then combat timing
+        // attacks by creating an arbitrary hash (should take about as long has comparing an existing one)
```

Comment:
```suggestion
        // attacks by creating an arbitrary hash (should take about as long as comparing an existing one)
```

Reply:
comparing? Creating?

Reply:
And I think it just more simple to use fake instead of arbitrary here :-) just more simple this way and easier to understand

Reply:
After `* * * Next comes the original logic * * *` I didn't change anything, everything is as in the original :)